src/ir/subtypes.h

```diff
@@ -73,6 +73,50 @@ struct SubTypes {
     return ret;
   }
 
+  // Computes the depth of children for each type. This is 0 if the type has no
+  // subtypes, 1 if it has subtypes but none of those have subtypes themselves,
+  // and so forth.
+  //
+  // This depth ignores bottom types.
+  std::unordered_map<HeapType, Index> getMaxDepths() {
+    std::unordered_map<HeapType, Index> depths;
+
+    // Begin with depth 0.
+    for (auto type : types) {
+      depths[type] = 0;
+    }
+
+    // Begin with a plan to work on all the types. When we visit an item, we'll
+    // update our super type based on our current depth.
+    std::unordered_set<HeapType> work(types.begin(), types.end());
```

Comment:
This algorithm is O(n^2), but I would expect that the other O(n^2) algorithm using two nested for loops and no work list would perform better.

An O(n) algorithm would be to topologically sort the types then collect the max depths in a single pass. We have a topo sort utility here: https://github.com/WebAssembly/binaryen/blob/main/src/support/topological_sort.h. You can see how it is used here: https://github.com/WebAssembly/binaryen/blob/main/src/support/topological_sort.h.

Reply:
Good idea! Done.